scheduler/server/load_based_sched_alg.go

```diff
@@ -0,0 +1,781 @@
+var (
+	DefaultLoadBasedSchedulerClassPercents = map[string]int32{"c0": 100}
+	DefaultRequestorToClassMap             = map[string]string{".*": "c0"}
+	DefaultMinRebalanceTime                = time.Duration(4 * time.Minute)
+	MaxTaskDuration                        = time.Duration(4 * time.Hour)
+)
+
+type LoadBasedAlgConfig struct {
+	classLoadPercents       map[string]int
+	classLoadPercentsMu     sync.RWMutex
+	requestorReToClassMap   map[string]string
+	requestorReToClassMapMU sync.RWMutex
+
+	rebalanceThreshold     int
+	rebalanceThresholdMu   sync.RWMutex
+	rebalanceMinDuration   time.Duration
+	rebalanceMinDurationMu sync.RWMutex
+
+	classByDescLoadPct []string
+
+	stat stats.StatsReceiver
+}
+
+// LoadBasedAlg the scheduling algorithm allocates job tasks to workers using a class map as
+// follows:
+// - each job maps to a 'class' (based on the job's requestor value)
+// - classes are assigned a % of the number of scoot workers.
+// When the algorithm is assigning tasks to workers it will try to start tasks such that the
+// number of running tasks maintain the defined class %'s.  We refer to number of workers as
+// per a class's defined % as the number of 'entitled' workers for the class.
+// (The class is entitled to use class % * total number of workers to run job tasks from jobs
+// assigned to the class.)
+// When there are not enough tasks to use all of the class's entitled workers, the algorithm
+// will allow other classes to run their tasks on the unused workers.  We refer to
+// this as loaning worker to other classes.
+// Note: workers are not assigned to specific classes.  The class % concept is simply a counting
+// mechanism.
+//
+// Each scheduling iteration tries to bring the task allocation back to the original class
+// entitlement as defined in the class %s. It could be the case that long running tasks slowly create
+// an imbalance in the worker to class numbers (long running tasks accumulate loaned workers).
+// As such, the algorithm periodically rebalances the running workers back toward the original target
+// %s by stopping tasks that have been started on loaned workers.  It will stop the most recently started
+// tasks till the running task to class allocation meets the original entitlement targets.
+//
+type LoadBasedAlg struct {
+	config     *LoadBasedAlgConfig
+	jobClasses map[string]*jobClass
+
+	totalUnusedEntitlement          int
+	exceededRebalanceThresholdStart time.Time
+
+	// local copy of load pcts and requestor map to use during assignment computation
+	// to insulate the computation from external changes to the configuration
+	classLoadPercents     map[string]int
+	requestorReToClassMap map[string]string
+
+	classByDescLoadPct             []string
+	tasksByJobClassAndStartTimeSec map[taskClassAndStartKey]taskStateByJobIDTaskID
+}
+
+// NewLoadBasedAlg allocate a new LoadBaseSchedAlg object.  If the load %'s don't add up to 100
```

Comment:
Doesn't seem to return errors or care what the %s add up to

Reply:
it normalizes the %s automatically

Reply:
I mean - the function is not doing what the comment is saying